File: mobile_app/griham_app/lib/models/bank_account.dart
class BankAccount {
  final String id;
  final String bankName;
  final String accountNumber;
  final String accountHolderName;
  final String accountName;
  final String type; // SAVINGS, CURRENT, CREDIT_CARD
  final double balance;
  final String currency;
  final bool isActive;

  BankAccount({
    required this.id,
    required this.bankName,
    required this.accountNumber,
    required this.accountHolderName,
    required this.accountName,
    required this.type,
    required this.balance,
    required this.currency,
    this.isActive = true,
  });

  factory BankAccount.fromJson(Map<String, dynamic> json) {
    return BankAccount(
      id: json['id'] ?? '',
      bankName: json['bank_name'] ?? '',
      accountNumber: json['account_number'] ?? '',
      accountHolderName: json['account_holder_name'] ?? '',
      accountName: json['account_name'] ?? json['bank_name'] ?? '',
      type: json['type'] ?? 'SAVINGS',
      balance: double.tryParse(json['balance']?.toString() ?? '0.0') ?? 0.0,
      currency: json['currency'] ?? '₹',
      isActive: json['is_active'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bank_name': bankName,
      'account_number': accountNumber,
      'account_holder_name': accountHolderName,
      'account_name': accountName,
      'type': type,
      'balance': balance,
      'currency': currency,
      'is_active': isActive,
    };
  }
}

File: microservice/src/db/schemas/BankAccount.ts
import mongoose, { Schema, Model } from 'mongoose';
import mongooseLeanVirtuals from 'mongoose-lean-virtuals';

export interface IBankAccountDoc {
  _id: string;
  id?: string;
  family_id: string;
  account_name: string;
  account_number: string | null;
  bank_name: string;
  account_type: string;
  balance: number;
  currency: string;
  created_at?: Date;
  updated_at?: Date;
}

const bankAccountSchema = new Schema<IBankAccountDoc>(
  {
    _id: { type: String, required: true },
    family_id: { type: String, required: true, ref: 'Family' },
    account_name: { type: String, required: true },
    account_number: { type: String, default: null },
    bank_name: { type: String, required: true },
    account_type: { type: String, default: 'savings' },
    balance: { type: Number, default: 0 },
    currency: { type: String, default: 'INR' },
  },
  { timestamps: true, id: false }
);

bankAccountSchema.index({ family_id: 1 });


bankAccountSchema.virtual('id').get(function () {
  return this._id;
});

bankAccountSchema.plugin(mongooseLeanVirtuals);

export const BankAccountModel: Model<IBankAccountDoc> = mongoose.model<IBankAccountDoc>(
  'BankAccount',
  bankAccountSchema
);
